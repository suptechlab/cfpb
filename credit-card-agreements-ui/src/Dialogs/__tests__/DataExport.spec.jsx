import { shallow } from 'enzyme'
import React from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import ReduxDataExport, { DataExport, mapDispatchToProps } from '../DataExport'

function setupEnzyme() {
  const props = {
    allComplaints: 999,
    exportAll: jest.fn(),
    exportSome: jest.fn(),
    onOtherFormats: jest.fn(),
    someComplaints: 99
  }

  const target = shallow(<DataExport {...props} />)

  return {
    props,
    target
  }
}

function setupSnapshot(total=1001) {
  const middlewares = [thunk]
  const mockStore = configureMockStore(middlewares)
  const store = mockStore({
    results: {
      doc_count: 9999,
      total
    },
  })

  return renderer.create(
    <Provider store={store}>
      <IntlProvider locale="en">
        <ReduxDataExport />
      </IntlProvider>
    </Provider>
  )
}

describe('component::DataExport', () => {
  describe('initial state', () => {
    it('renders without crashing', () => {
      const target = setupSnapshot()
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('hides the dataset radio buttons when there is no filter', () => {
      const target = setupSnapshot(9999)
      const tree = target.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('user interaction', () => {
    let target, props
    beforeEach(() => {
      ({target, props} = setupEnzyme())
    })

    it('supports changing the format', () => {
      const radio = target.find('#format_csv')
      radio.simulate('change', { target: radio.getElements()[0].props })
      expect(target.state('format')).toEqual('csv')
    })

    it('supports changing the which dataset is used', () => {
      const radio = target.find('#dataset_filtered')
      radio.simulate('change', { target: radio.getElements()[0].props })
      expect(target.state('dataset')).toEqual('filtered')
    })

    it('provides link to Socrata', () => {
      const btn = target.find('.other-formats button')
      btn.simulate('click')
      expect(props.onOtherFormats).toHaveBeenCalled()
    })

    describe('clicking Start Export', () => {
      let startExport
      beforeEach(() => {
        startExport = target.find('.footer button').first()
      })

      it('is disabled at the beginning', () => {
        expect(startExport.prop('disabled')).toEqual(true)
      })

      it('calls a specific action when "filtered" is chosen', () => {
        target.setState({dataset: 'filtered'})
        startExport.simulate('click')
        expect(props.exportSome).toHaveBeenCalled()
      })

      it('calls a specific action when "all" is chosen', () => {
        target.setState({dataset: 'full'})
        startExport.simulate('click')
        expect(props.exportAll).toHaveBeenCalled()
      })

      it('displays the long download warning', () => {
        target.setState({dataset: 'full'})
        startExport.simulate('click')
        expect(target.state('mode')).toEqual('NOTIFYING')
      })
    })
  })

  describe('componentWillReceiveProps', () => {
    it('reruns all the validations', () => {
      const {target} = setupEnzyme()
      target.setProps({someComplaints: 999})
      expect(target.state('messages')).toEqual(
        { format: 'You must select a format for the export' }
      )
    })
    
  })

  describe('mapDispatchToProps', () => {
    it('provides a way to call visitSocrata', () => {
      const dispatch = jest.fn()
      mapDispatchToProps(dispatch).onOtherFormats()
      expect(dispatch.mock.calls.length).toEqual(1)
    })

    it('provides a way to call exportAllResults', () => {
      const dispatch = jest.fn()
      mapDispatchToProps(dispatch).exportAll('json')
      expect(dispatch.mock.calls.length).toEqual(1)
    })

    it('provides a way to call exportSomeResults', () => {
      const dispatch = jest.fn()
      mapDispatchToProps(dispatch).exportSome('csv', 1300)
      expect(dispatch.mock.calls.length).toEqual(1)
    })
  })
})
