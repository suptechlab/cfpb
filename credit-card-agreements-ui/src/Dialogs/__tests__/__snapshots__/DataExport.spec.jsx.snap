// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::DataExport initial state hides the dataset radio buttons when there is no filter 1`] = `
<section
  className="export-modal"
>
  <div
    className="header layout-row"
  >
    <h3
      className="flex-all"
    >
      Export complaints
    </h3>
    <button
      className="a-btn a-btn__link"
      data-gtm_ignore="true"
      onClick={undefined}
    >
      Close
      <span
        className="cf-icon cf-icon-delete-round"
      />
    </button>
  </div>
  <div
    className="body"
  >
    <div
      className="body-copy instructions"
    >
      To download a copy of this dataset, choose the file format and which complaints you want to export below.
    </div>
    <div
      className="group"
    >
      <div
        className="group-title"
      >
        Select a format for the exported file
        <span
          className="required-helper"
        >
           (required)
        </span>
      </div>
      <div
        className="body-copy"
      >
        <div
          className="m-form-field m-form-field__radio m-form-field__lg-target"
        >
          <input
            checked={false}
            className="a-radio"
            id="format_csv"
            onChange={[Function]}
            type="radio"
            value="csv"
          />
          <label
            className="a-label"
            htmlFor="format_csv"
          >
            CSV
          </label>
        </div>
        <div
          className="m-form-field m-form-field__radio m-form-field__lg-target"
        >
          <input
            checked={false}
            className="a-radio"
            id="format_json"
            onChange={[Function]}
            type="radio"
            value="json"
          />
          <label
            className="a-label"
            htmlFor="format_json"
          >
            JSON
          </label>
        </div>
      </div>
      <div
        className="other-formats"
      >
        Or you can 
        <button
          className="a-btn a-btn__link"
          data-gtm_ignore="true"
          onClick={[Function]}
        >
          download the data in a different format
        </button>
        , if needed.
      </div>
    </div>
    <div
      className="timeliness-warning"
    >
      The export process could take several minutes if you're downloading many complaints
    </div>
  </div>
  <div
    className="footer layout-row"
  >
    <button
      className="a-btn"
      data-gtm_ignore="true"
      disabled={true}
      onClick={[Function]}
    >
      Start Export
    </button>
    <button
      className="a-btn a-btn__link a-btn__warning"
      data-gtm_ignore="true"
      onClick={undefined}
    >
      Cancel
    </button>
  </div>
</section>
`;

exports[`component::DataExport initial state renders without crashing 1`] = `
<section
  className="export-modal"
>
  <div
    className="header layout-row"
  >
    <h3
      className="flex-all"
    >
      Export complaints
    </h3>
    <button
      className="a-btn a-btn__link"
      data-gtm_ignore="true"
      onClick={undefined}
    >
      Close
      <span
        className="cf-icon cf-icon-delete-round"
      />
    </button>
  </div>
  <div
    className="body"
  >
    <div
      className="body-copy instructions"
    >
      To download a copy of this dataset, choose the file format and which complaints you want to export below.
    </div>
    <div
      className="group"
    >
      <div
        className="group-title"
      >
        Select a format for the exported file
        <span
          className="required-helper"
        >
           (required)
        </span>
      </div>
      <div
        className="body-copy"
      >
        <div
          className="m-form-field m-form-field__radio m-form-field__lg-target"
        >
          <input
            checked={false}
            className="a-radio"
            id="format_csv"
            onChange={[Function]}
            type="radio"
            value="csv"
          />
          <label
            className="a-label"
            htmlFor="format_csv"
          >
            CSV
          </label>
        </div>
        <div
          className="m-form-field m-form-field__radio m-form-field__lg-target"
        >
          <input
            checked={false}
            className="a-radio"
            id="format_json"
            onChange={[Function]}
            type="radio"
            value="json"
          />
          <label
            className="a-label"
            htmlFor="format_json"
          >
            JSON
          </label>
        </div>
      </div>
      <div
        className="other-formats"
      >
        Or you can 
        <button
          className="a-btn a-btn__link"
          data-gtm_ignore="true"
          onClick={[Function]}
        >
          download the data in a different format
        </button>
        , if needed.
      </div>
    </div>
    <div
      className="group"
    >
      <div
        className="group-title"
      >
        Select which complaints you'd like to export
        <span
          className="required-helper"
        >
           (required)
        </span>
      </div>
      <div
        className="body-copy"
      >
        <div
          className="m-form-field m-form-field__radio m-form-field__lg-target"
        >
          <input
            checked={false}
            className="a-radio"
            id="dataset_filtered"
            onChange={[Function]}
            type="radio"
            value="filtered"
          />
          <label
            className="a-label"
            htmlFor="dataset_filtered"
          >
            <div
              className="multiline-label"
            >
              <div>
                Filtered dataset (
                <span>
                  1,001
                </span>
                 complaints)
              </div>
              <div
                className="body-copy"
              >
                (only the results of the last search and/or filter)
              </div>
            </div>
          </label>
        </div>
        <div
          className="m-form-field m-form-field__radio m-form-field__lg-target"
        >
          <input
            checked={false}
            className="a-radio"
            id="dataset_full"
            onChange={[Function]}
            type="radio"
            value="full"
          />
          <label
            className="a-label"
            htmlFor="dataset_full"
          >
            <div
              className="multiline-label"
            >
              <div>
                Full dataset (
                <span>
                  9,999
                </span>
                 complaints)
              </div>
              <div
                className="body-copy"
              >
                (not recommended due to very large file size)
              </div>
            </div>
          </label>
        </div>
      </div>
    </div>
    <div
      className="timeliness-warning"
    >
      The export process could take several minutes if you're downloading many complaints
    </div>
  </div>
  <div
    className="footer layout-row"
  >
    <button
      className="a-btn"
      data-gtm_ignore="true"
      disabled={true}
      onClick={[Function]}
    >
      Start Export
    </button>
    <button
      className="a-btn a-btn__link a-btn__warning"
      data-gtm_ignore="true"
      onClick={undefined}
    >
      Cancel
    </button>
  </div>
</section>
`;
